### ISSUE TITLE
Spike: how to export CSV from VES

### ISSUE THREAD

**Issue by wesrowe** [department-of-veterans-affairs/va.gov-cms#13759]
### User story

**AS A** PW Engineer
**I WANT** to figure out how to export ONE of the Income Limits tables in VES to CSV
**SO THAT** I feel confident we can manage and automate our own exports.

### Engineering notes / background

Why a spike:
- This is the first time we've live-queried the VES database
- This is written in Ruby because the subnet that we need to leverage is between vets-api (Ruby) and Oracle
  - In the past, Joshua Faulkner (of VES) exported the tables from within Oracle environment
- Development approach: we can't run Oracle locally alongside vets-api, so process will be very awkward and time-intensive
- Ruby will need a bundle/gem/something to connect to Oracle (?)
  - It will be an ORM-style query that an Oracle SME can't help with; but query is pretty simple (select all from...)

Ideas of where to store CSV as this intermediate step:
- Add endpoints for queries

## Acceptance criteria

Timebox: 8, come up for air and discuss, assess

- [ ] Figure out where this CSV can be sent to so we can see it...
- [ ] Script exports one of the VES tables we need to CSV to...? [This AC is risky and may not fit in the 8 points / may need to be carved into a separate ticket]
    - [ ] Diff between new and existing (on prod S3) CSVs shows that the new one is the same [Dependent on the script]

**Comment by wesrowe** [comment]
Discussion in planning: this could be a very deep can of worms; ACs are aggressive; committing to the timebox only, will see where we get then reassess.

**Comment by jilladams** [comment]
Daniel flagged may have a dependency we should ticket / prioritize first, Daniel will write the ticket.

**Comment by jilladams** [comment]
Tom Greene at A6 eng coach can provide support if you need for Ruby stuff.

**Comment by dsasser** [comment]
# Mid Sprint Update 9/5/23

🔴 Blocked, but working to unblock with Platform.

In order to connect vets-api to Oracle, the `ruby-oci8` package is needed. Unfortunately, this package is not compatible with Apple Silicon chips because it depends on an Oracle Client library which has not released a version compatible with the most modern Macs.

![Screenshot 2023-09-05 at 10 47 37 AM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/221539/364b2ee6-17a5-46d1-9755-91cadc9da24b)

This means that no local development can be done natively on OSX with Apple chips.

While it may be possible to use Docker for all development, simply having the `ruby-oci8` package as a requirement for vets-api will impact all vets-api developers who are using the native or hybrid setup (which, anecdotally, is most vets-api devs). Using Docker then would unblock me, but affect all other vets-api devs going forward. This is a non-starter.

I have [approached Platform](https://dsva.slack.com/archives/CBU0KDSB1/p1693933955386729) with this and am working with them to devise a solution.

**Comment by dsasser** [comment]
# Update

Because Ruby is so far a non-starter, I've been thinking of alternatives to using vets-api to perform this task. After all, we don't need vets-api to do this per say, so long as it is done.

I pondered if Github Actions would work, and Platform folks suggested this to be a viable alternative. I [approached CMS DevOps with the idea](https://dsva.slack.com/archives/CT4GZBM8F/p1693936884434949), and Nate and Tim provided some excellent information on getting a private runner setup.

Using a private runner would afford us the option to use other programming languages such as Node, which don't have the CPU limitation. Or, we could still write in Ruby if we chose to do that.

Private runners run inside the VA network, rather than on Github's network, so in theory they should be able to access the Oracle db. And, even if they can't access the DB today, we can always go through ESECC to get that setup just as we did for vets-api.

**Comment by dsasser** [comment]
# SPIKE Results

## Blocked on Ruby
We originally planned for this work to live in the vets-api codebase (Ruby), but ran into blockers noted in [earlier](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/13759#issuecomment-1707047917) comments. To summarize, the Oracle client library is not compatible with modern Apple silicon CPUs because Apple changed from x86-64 architecture to ARM64. Oracle has not updated their client library to work with ARM64 architectures, which is an internal component to the ruby-oci8 package that runs Oracle client library for Ruby.

## Alternatives
In thinking about alternatives, I wanted to break down to the core of what immutable needs exist to accomplish a CSV export from VES: 

- a machine compatible with Oracle client libraries
- ability to write to S3
- network connectivity to VES

This led me to consider [Github self-hosted runners](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners). This approach nearly immediately meets all our needs:

- 🟢 a machine compatible with Oracle client libraries
- 🟢 ability to write to S3
- 🟡 network connectivity to VES

The network connectivity remains untested. However, even if not immediately functional, we can always go through ESECC to establish the network path just as we did for vets-api.

## Next Steps

### Run a test:
Create a new self-hosted runner likely using the method described in [Tim's comment](https://dsva.slack.com/archives/CT4GZBM8F/p1693941770594819?thread_ts=1693936884.434949&cid=CT4GZBM8F).

The runner should execute a java test script created by Johnathan Falkner for the purposes of testing network connectivity to the VES network.

The test should execute the following commands:
- copy the connection test from s3: `aws s3 cp s3://dsva-vagov-staging-config-bucket/cms-test/ConnectionTest.jar . --region us-gov-west-1`
- install java: `yum install java-1.8.0-openjdk`
- run the test: `java -jar ConnectionTest.jar`

Should the test prove unsuccessful, we will need to work with ESECC/Gateway Operations to establish connectivity.

### Start Developing the CSV Export
Once network connectivity is established, we can move to development.

Some supported languages:

- PHP (OCI8)
- Python (cx_Oracle), 
- Node.js (node-oracledb)
- Ruby (ruby-oci8)

I suggest Node or PHP, since PW developers know and can support those languages natively.

**Comment by jilladams** [comment]
From 1:1 with Daniel today: He'll cut tickets to capture the next steps above, and this ticket can close. We didn't end up with a Script out of this spike, due to the env / Mac OS issues, but that will happen as next steps.
